controllers/accountclaim/accountclaim_controller.go

```diff
@@ -231,6 +393,181 @@ func (r *AccountClaimReconciler) Reconcile(ctx context.Context, request ctrl.Req
+	return &corev1.Secret{
+		Type: "Opaque",
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: secretNameSpace,
+		},
+		Data: map[string][]byte{
+			"role_arn": arn,
+		},
+	}
+
+}
+
+// CreateOrUpdateSecret creates a secret in AWS Secrets Manager or updates it if it already exists.
+func (r *AccountClaimReconciler) createIAMRoleSecret(reqLogger logr.Logger, accountClaim *awsv1alpha1.AccountClaim, roleARN string) error {
+	var OCMSecretNamespace string
+
+	OCMSecretName := awsSTSSecret
+	if accountClaim.Spec.AwsCredentialSecret.Namespace == "" {
+		OCMSecretNamespace = accountClaim.ObjectMeta.Namespace
+
+	} else {
+		OCMSecretNamespace = accountClaim.Spec.AwsCredentialSecret.Namespace
+	}
+
+	OCMSecret := newStsSecretforCR(OCMSecretName, OCMSecretNamespace, []byte(roleARN))
+
+	err := r.Client.Create(context.TODO(), OCMSecret)
+	if err != nil {
+		reqLogger.Error(err, "Unable to create secret for OCM")
+		return err
+	}
+	reqLogger.Info(fmt.Sprintf("Secret %s created for claim %s", OCMSecret.Name, accountClaim.Name))
+
+	accountClaim.Spec.AwsCredentialSecret.Name = OCMSecretName
+	err = r.Client.Update(context.TODO(), accountClaim)
+	if err != nil {
+		reqLogger.Error(err, fmt.Sprintf("AccountClaim spec update for %s failed", accountClaim.Name))
+	}
+	return nil
+}
+
+// CreateIAMRoleWithPermissions creates an IAM role with the specified permissions' policy.
+func (r *AccountClaimReconciler) createIAMRoleWithPermissions(reqLogger logr.Logger, awsClient awsclient.Client, roleName string, trustedARN string) (string, error) {
+	type awsStatement struct {
+		Effect    string                 `json:"Effect"`
+		Action    []string               `json:"Action"`
+		Resource  []string               `json:"Resource,omitempty"`
+		Principal *awsv1alpha1.Principal `json:"Principal,omitempty"`
+	}
+
+	assumeRolePolicyDoc := struct {
+		Version   string
+		Statement []awsStatement
+	}{
+		Version: "2012-10-17",
+		Statement: []awsStatement{{
+			Effect: "Allow",
+			Action: []string{"sts:AssumeRole"},
+			Principal: &awsv1alpha1.Principal{
+				AWS: []string{trustedARN},
+			},
+		}},
+	}
+	// Convert role to JSON
+	jsonAssumeRolePolicyDoc, err := json.Marshal(assumeRolePolicyDoc)
+	if err != nil {
+		return "", err
+	}
+
+	reqLogger.Info(fmt.Sprintf("Creating role: %s", roleName))
+	createRoleOutput, err := awsClient.CreateRole(&iam.CreateRoleInput{
+		RoleName:                 aws.String(roleName),
+		Description:              aws.String(""), //description
+		AssumeRolePolicyDocument: aws.String(string(jsonAssumeRolePolicyDoc)),
+	})
+	if err != nil {
+		return "", err
+	}
+	reqLogger.Info(fmt.Sprintf("Role %s created", createRoleOutput))
+
+	arnComponents := strings.Split(*createRoleOutput.Role.Arn, ":")
+
+	policyDocument, err := generateInlinePolicy(arnComponents[4])
```

Comment:
I think [4] is the `account`, so maybe assign it to a variable named accountid, to make it more clear.